### PR TITLE
[Relayminer] Fix: relayminer on time sessions

### PR DESCRIPTION
**Pull request by bryanchriswhite**
## Summary

Re-writes `mapBlockToSessionsToClaim()` for use in a `channel.ForEach` (as opposed to the `channel.Map`) to ensure that late & on-time sessiontrees are emitted without an undesired delay in-between.

## Issue

image
> See: https://github.com/pokt-network/poktroll/pull/546#discussion_r1613393041

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: pkg/relayer/session/session.go [hunk 1]
```diff
@@ -163,92 +165,91 @@ func (rs *relayerSessionsManager) ensureSessionTree(sessionHeader *sessiontypes.
 	return sessionTree, nil
 }
 
-// mapBlockToSessionsToClaim is a MapFn which maps committed blocks to a list of
-// sessions which can be claimed as of that block.
+// mapBlockToSessionsToClaimFn returns a new MapFn which maps committed blocks to
+// a list of sessions which can be claimed as of that block.
```
Comment (bryanchriswhite) [inline pkg/relayer/session/session.go:169]:
```suggestion
// forEachBlockClaimSessionsFn returns a new ForEachFn that sends a lists of sessions
// which are eligible to be claimed at each block height, effectively mapping committed
// blocks to a list of sessions which can be claimed as of that block.
// If "late" sessions are found, they are emitted as quickly as possible & are expected
// to bypass downstream delay operations. "late" sessions are emitted, as they're discovered
// (by iterating over map keys), as a distinct notifications from "on-time" sessions & other
// "late" sessions. Under nominal conditions, only one set of "on-time" sessions (w/ the same
// session start/end heights) should be present in the rs.sessionsTrees map. "Late" sessions
// are expected to present in the presence of network interruptions, restarts, or other
// disruptions to the relayminer process.
```

File: pkg/relayer/session/session.go [hunk 2]
```diff
@@ -163,92 +165,91 @@ func (rs *relayerSessionsManager) ensureSessionTree(sessionHeader *sessiontypes.
...
-	// They will be emitted last, after all the late sessions have been emitted.
-	var onTimeSessions []relayer.SessionTree
-
-	// TODO_IMPROVE(#543): Params are currently queried each block.
-	sharedParams, err := rs.sharedQueryClient.GetParams(ctx)
-	if err != nil {
-		rs.logger.Error().Err(err).Msg("unable to query shared module params")
-		return nil, true
-	}
+func (rs *relayerSessionsManager) mapBlockToSessionsToClaimFn(
```
Comment (bryanchriswhite) [inline pkg/relayer/session/session.go:172]:
```suggestion
func (rs *relayerSessionsManager) mapBlockToSessionsToClaimFn(
```

File: pkg/relayer/session/session.go [hunk 3]
```diff
@@ -89,10 +89,12 @@ func NewRelayerSessions(
...
 	}
 
-	rs.sessionsToClaimObs = channel.Map(
+	sessionsToClaimObs, sessionsToClaimPublishCh := channel.NewObservable[[]relayer.SessionTree]()
+	rs.sessionsToClaimObs = sessionsToClaimObs
+	channel.ForEach(
 		ctx,
 		rs.blockClient.CommittedBlocksSequence(ctx),
-		rs.mapBlockToSessionsToClaim,
+		rs.mapBlockToSessionsToClaimFn(sessionsToClaimPublishCh),
```
Comment (bryanchriswhite) [inline pkg/relayer/session/session.go:97]:
```suggestion
		rs.forEachBlockClaimSessionsFn(sessionsToClaimPublishCh),
```

File: pkg/relayer/session/session.go [hunk 4]
```diff
@@ -163,92 +165,99 @@ func (rs *relayerSessionsManager) ensureSessionTree(sessionHeader *sessiontypes.
 	return sessionTree, nil
 }
 
-// mapBlockToSessionsToClaim is a MapFn which maps committed blocks to a list of
-// sessions which can be claimed as of that block.
+// forEachBlockClaimSessionsFn returns a new ForEachFn that sends a lists of sessions which
+// are eligible to be claimed at each block height on sessionsToClaimsPublishCh, effectively
+// mapping committed blocks to a list of sessions which can be claimed as of that block.
+// If "late" sessions are found, they are emitted as quickly as possible & are expected
```
Comment (Olshansk) [inline pkg/relayer/session/session.go:171]:
```suggestion
// If "late" sessions are found, they are emitted as quickly as possible and are expected
```

File: pkg/relayer/session/session.go [hunk 5]
```diff
@@ -163,92 +165,99 @@ func (rs *relayerSessionsManager) ensureSessionTree(sessionHeader *sessiontypes.
 	return sessionTree, nil
 }
 
-// mapBlockToSessionsToClaim is a MapFn which maps committed blocks to a list of
-// sessions which can be claimed as of that block.
+// forEachBlockClaimSessionsFn returns a new ForEachFn that sends a lists of sessions which
```
Comment (Olshansk) [inline pkg/relayer/session/session.go:168]: Can you try to rework this to have bullet points. It's a great comment, which makes sense, but really took a lot of cognitive effort to read & undersand.

**Review by Olshansk — APPROVED** [review]
Approving but also making one small request to the comments.

**Review by red-0ne — APPROVED** [review]
LGTM. Thanks for the detailed description of `forEachBlockClaimSessionsFn`

**Pull request merged**